### PR TITLE
Fixed dice-genetic efficiency bugs - normalized losses and fixed predict function bug

### PR DESCRIPTION
**Pull request by soundarya98**
In response to #150: 

Normalized losses so that losses don't dominate over each other. Before, proximity and sparsity losses were dominating the overall loss, and we were exiting the loop only when the cfs were in the desired_class, which led to arbitrarily long runs sometimes
- Normalized Sparsity loss by dividing by the number of features
- Normalized Proximity loss by dividing by the sum of feature weights

Also fixed a bug wrt argmax and compute_yloss and desired_class 2 - 
Earlier, for the predict scores [0, 0.5, 0,5], yloss was 0 as class 2 had the same value as the maximum score, but argmax showed class 1 instead of 2. 
- Fixed this by including a custom predict function that returns the desired_class if the score[desired_class] = maxvalue, and argmax otherwise

### CODE REVIEW

**Review by gaugup — COMMENTED** [review]
Could you please add tests for the mentioned bugs that you are fixing in this review?

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 1]
```diff
@@ -306,6 +309,20 @@ def predict_fn(self, input_instance):
         input_instance = self.label_decode(input_instance)
         return self.model.get_output(input_instance, model_score=False)
 
+    def predict_fn_custom(self, input_instance, desired_class):
```
[the remark this answers is not in the document]
Comment (soundarya98) [reply dice_ml/explainer_interfaces/dice_genetic.py:312]: I couldn't call this function from the test module if it was private, so I've included a single underscore before the function

**Comment by soundarya98** [comment]
> Could you please add tests for the mentioned bugs that you are fixing in this review?

**Review by gaugup — APPROVED** [review]

**Comment by gaugup** [comment]
> Could you please add tests for the mentioned bugs that you are fixing in this review?

@soundarya98 thanks for adding the unit tests. Changes look good. Approved it.

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 2]
```diff
@@ -306,6 +309,20 @@ def predict_fn(self, input_instance):
         input_instance = self.label_decode(input_instance)
         return self.model.get_output(input_instance, model_score=False)
 
+    def _predict_fn_custom(self, input_instance, desired_class):
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:312]: can you write some comments for the logic here? saying that why we should prefer the output to be `desired_class`, and we actually change it. (and why it is needed--the issue with sklearn's predict)

File: dice_ml/explainer_interfaces/dice_pytorch.py [hunk 3]
```diff
@@ -376,8 +376,8 @@ def stop_loop(self, itr, loss_diff):
         """Determines the stopping condition for gradient descent."""
 
         # intermediate projections
-        if((self.project_iter > 0) & (itr > 0)):
-            if((itr % self.project_iter) == 0):
+        if self.project_iter > 0 and itr > 0:
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_pytorch.py:379]: why are all these unrelated changes part of this PR? I see a lot of formatting changes for if.

File: tests/test_dice_interface/test_dice_genetic.py [hunk 4]
```diff
@@ -140,6 +140,15 @@ def test_zero_cfs(self, desired_class, sample_custom_query_2, total_CFs, initial
                                            total_CFs=total_CFs, desired_class=desired_class,
                                            initialization=initialization)
 
+    # Testing the custom predict function
+    @pytest.mark.parametrize("desired_class", [2])
+    def test_predict_custom(self, desired_class, sample_custom_query_2, mocker):
```
Comment (amit-sharma) [inline tests/test_dice_interface/test_dice_genetic.py:145]: like the mocker approach.

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 5]
```diff
@@ -306,6 +309,20 @@ def predict_fn(self, input_instance):
...
+        """Checks that the maximum predicted score lies in the desired class"""
+        input_instance = self.label_decode(input_instance)
+        output = self.model.get_output(input_instance, model_score=True)
+        desired_class = int(desired_class)
+        maxvalues = np.max(output, 1)
+        predicted_values = np.argmax(output, 1)
+
+        for i in range(len(output)):
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:320]: inline comment for this line will be useful

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 6]
```diff
@@ -325,30 +342,34 @@ def compute_yloss(self, cfs, desired_range, desired_class):
...
+        feature_weights = np.array(
+            [self.feature_weights_list[0][i] for i in self.data_interface.continuous_feature_indexes])
+        product = np.multiply(
+            (abs(x_hat - query_instance_normalized)[:, [self.data_interface.continuous_feature_indexes]]),
+            feature_weights)
         product = product.reshape(-1, product.shape[-1])
-        proximity_loss = np.sum(product, axis=1)
-        return proximity_loss
+        proximity_loss = np.sum(product,
+                                axis=1)  # Dividing by the sum of feature weights to normalize proximity loss
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:359]: comment at end of line 359 should be on a new line.

File: dice_ml/explainer_interfaces/dice_genetic.py [hunk 7]
```diff
@@ -325,30 +342,34 @@ def compute_yloss(self, cfs, desired_range, desired_class):
...
+        return proximity_loss / sum(feature_weights)
 
     def compute_sparsity_loss(self, cfs):
         """Compute weighted distance between two vectors."""
         sparsity_loss = np.count_nonzero(cfs - self.x1, axis=1)
-        return sparsity_loss
+        return sparsity_loss / len(
+            self.data_interface.feature_names)  # Dividing by the number of features to normalize sparsity loss
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_genetic.py:366]: should the sparsity_loss be divided by total number of features or the total number of one-hot encoded features? Good to check here--I'm not sure.

**Review by amit-sharma — COMMENTED** [review]
Added comments.  Two questions:
1. Do these changes reduce the arbitrary running time for some examples? 
2. Why are many formatting changes (especially with if statements) a part of this PR?

File: dice_ml/explainer_interfaces/dice_genetic.py (hunk 7 above)
Comment (soundarya98) [reply dice_ml/explainer_interfaces/dice_genetic.py:366]: Great point!

Comment (soundarya98) [reply dice_ml/explainer_interfaces/dice_genetic.py:366]: It will always be the number of features since the compute_sparsity_loss is only called for members of the population which are constructed initially in the following statement:
`self.cfs = np.zeros((self.population_size, self.data_interface.number_of_features))`

File: dice_ml/explainer_interfaces/dice_pytorch.py (hunk 3 above)
Comment (soundarya98) [reply dice_ml/explainer_interfaces/dice_pytorch.py:379]: I was trying to fix all if (X) to if X at once and that's why it all ended up in the same PR. Shifting it to a different PR

**Review by amit-sharma — APPROVED** [review]
